### ISSUE TITLE
[question/feature] conanfile environment include/exclude env settings from certain dependencies

### ISSUE THREAD

**Issue by Johnnyxy** [conan-io/conan#4601]
Hi there,
I have build dependencies that I need at certain points in the build process.

prerequisites
---------------
I have a build dependency (code generator; dep_a) which has env-values that interfere with env-values of another build dependency (dep_b). At conan-build I am executing CMake (with build helper). 

workflow
----------
At CMake's build execution I need to build a target with the env-values of the specific dependency (dep_a) and after that I need to build a CMake target completely without this dependency.
```python
def build(self):
    cmake = CMake(self)
    cmake.configure()
    cmake.build(target="target_with_dep_a")    
    cmake.build(target="target_without_dep_a")
```

question
----------
Is it possible to exclude the env-values of a specific build dependency?

workaround
-------------
I was digging in the conan code and I think it would be possible to do it the following way.
The code is taken and modified from docs [apply_env (step 1)](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#apply-env
) and conan code [env (step 2)](https://github.com/conan-io/conan/blob/68ced146342b04e56cd66fce729c86982198f471/conans/model/conan_file.py#L164):
```python
def build(self):
    cmake = CMake(self)
    cmake.configure()
    # ----------- step 1 -----------
    # apply normal deps_env_info
    with tools.environment_append(self.env):
        cmake.build(target="target_with_dep_a")
    
    # ----------- step 2 -----------
    # create a copy from normal deps_env_info
    # remove "dep_a" env_info and apply
    my_deps_env_info = copy.deepcopy(self.deps_env_info)
    del my_deps_env_info["dep_a"]

    tmp_env_values = self._conan_env_values.copy()
    tmp_env_values.update(my_deps_env_info)
    ret, multiple = tmp_env_values.env_dicts(self.name)
    ret.update(multiple)
    with tools.environment_append(ret):
        cmake.build(target="target_without_dep_a")
```

I did not actually test this as I am in the weekend at the moment :). I am going to test this tomorrow.

Best regards

**Comment by lasote** [comment]
If both build requires uses the same environment variables names maybe it is enough to apply a `tools.environment_append(self.deps_env_info["dep_a"]`  and then `tools.environment_append(self.deps_env_info["dep_b"]`.

Otherwise (not tested) could be to clear the environment from all the inherited values first, and then apply the individual ones:

```python
with tools.environment_append({k: None for k in self.env.keys()):
    with tools.environment_append(self.deps_env_info["dep_a"]):
          cmake.build(tagrget="target_without_b")
```

Let me know how it goes!

**Comment by Johnnyxy** [comment]
hm this would be much shorter :)

would this go too?

```python
with tools.environment_append(self.env):
    cmake.build(target="target_with_a")
with tools.environment_append({k: v for k, v in self.env if not key == "dep_a"):
    cmake.build(target="target_without_a")
```

**Comment by lasote** [comment]
No, the `self.env` doesn't have "dep_a" as a key but the whole env.
Let me know if you found already a fix.

**Comment by Johnnyxy** [comment]
hi, I will go for it in some time. But at the moment my workload is around 110% :).
As soon as I have some spare time I will try this and give you some feedback.

**Comment by Johnnyxy** [comment]
@lasote I was looking into the issue.
I have multiple dependencies which have their own enviroment settings.
Now I need to exclude one/or multiple dependencies from including their environment settings at certain points in the build process.

With your proposal it works with just one other dependency.

Workaround
-------------------
The following code is just a workaround and I would be pleased if this would work with Conan interface functions. It could be seen as hint how to solve the issue in one way, not neccessarily the best way.

What I did now was to make a copy of the [RunEnvironment](https://github.com/conan-io/conan/blob/release/1.13.0/conans/client/run_environment.py#L3) class and modified the input parameter to take a dictionary of dependencies instead of a conanfile-object.
With this approach I can selectively specify which dependencies gets appended to the environment at certain points in the build process.

The following code is written to use Python 3 (e.g. _dict.values()_):

```python
class DepsRunEnvironment(object):
    """
    - PATH: pointing to the bin/ directories of the requires
    - LD_LIBRARY_PATH: requires lib_paths for Linux
    - DYLD_LIBRARY_PATH: requires lib_paths for OSx
    """
    def __init__(self, deps_cpp_infos):
        """
        :param deps_cpp_infos: dict of dependencies cpp informations
        """
        self.deps_cpp_infos = deps_cpp_infos

    @property
    def vars(self):
        lib_paths = []
        bin_paths = []
        for v in self.deps_cpp_infos.values():
            lib_paths.extend(v.lib_paths)
            bin_paths.extend(v.bin_paths)

        ret = {"DYLD_LIBRARY_PATH": lib_paths,
               "LD_LIBRARY_PATH": lib_paths,
               "PATH": bin_paths}

        return ret
```

usage:

```python
def build(self):
    deps_cpp_infos = {k: self.deps_cpp_info[k] for k in list(self.deps_cpp_info.deps)}
    deps_cpp_infos_vars = DepsRunEnvironment(deps_cpp_infos).vars
        with tools.environment_append(deps_cpp_infos_vars):
            ...    

    deps_cpp_infos = {k: self.deps_cpp_info[k] for k in list(self.deps_cpp_info.deps) if not k == "dep_a"}
    deps_cpp_infos_vars = DepsRunEnvironment(deps_cpp_infos).vars
        with tools.environment_append(deps_cpp_infos_vars):
            ...    
```

The first DepsRunEnvironment block is just for clarification how to use this with all dependencies as it would work with ```RunEnvironment(self).vars``` too.

**Comment by memsharded** [comment]
Closing this ticket as outdated, both Conan 1 new ``Environment`` and Conan 2 have better control over env-vars, please check and report new issues against them if necessary, thanks.
